beginner_source/pendulum.py

```diff
@@ -0,0 +1,906 @@
+# Transforming an environment
+# ---------------------------
+#
+# Writing environment transforms for stateless simulators is slightly more
+# complicated than for stateful ones: transforming an output entry that needs
+# to be read at the following iteration requires to apply the inverse transform
+# before calling :func:`env.step` at the next step.
+# This is an ideal scenario to showcase all the features of torchrl's
+# transforms!
+#
+# For instance, in the following transformed environment we unsqueeze the entries
+# ``["th", "thdot"]`` to be able to stack them along the last
+# dimension. We also pass them as ``in_keys_inv`` to squeeze them back to their
+# original shape once they are passed as input in the next iteration.
+#
+env = TransformedEnv(
+    env,
+    # Unsqueezes the observations that we will concatenate
+    UnsqueezeTransform(
+        unsqueeze_dim=-1,
+        in_keys=["th", "thdot"],
+        in_keys_inv=["th", "thdot"],
+    ),
+)
+
+######################################################################
+# Writing custom transforms
+# ^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# TorchRL's transforms may not cover all the operations one wants to execute
+# after an environment has been executed.
+# Writing a transform does not require much effort. As for the environment
+# design, there are two steps in writing a transform:
+#
+# - Getting the dynamics right (forward and inverse);
+# - Adapting the environment specs.
+#
+# A transform can be used in two settings: on its own, it can be used as a
+# :class:`torch.nn.Module`. It can also be used appended to a
+# :class:`~torchrl.envs.TransformedEnv`. The structure of the class allows to
+# customize the behavior in the different contexts.
+#
+# A :class:`~torchrl.envs.Transform` skeleton can be summarized as follows:
+#
+# .. code-block::
+#
+#   class Transform(nn.Module):
+#       def forward(self, tensordict):
+#       def _apply_transform(self, tensordict):
+#       def _step(self, tensordict):
+#       def _call(self, tensordict):
+#       def inv(self, tensordict):
+#       def _inv_apply_transform(self, tensordict):
+#
+# There are three entry points (:func:`forward`, :func:`_step` and :func:`inv`)
+# which all receive :class:`tensordict.TensorDict` instances. The first two
+# will eventually go through the keys indicated by :obj:`Transform.in_keys`
+# and call :func:`Transform._apply_transform` to each of these. The results will
+# be written in the entries pointed by :obj:`Transform.out_keys` if provided
+# (if not the ``in_keys`` will be updated with the transformed values).
+# If inverse transforms need to be executed, a similar data flow will be
+# executed but with the :func:`Transform.inv` and
+# :func:`Transform._inv_apply_transform` methods and across the ``in_keys_inv``
+# and ``out_keys_inv`` list of keys.
+# The following figure summarized this flow for environments and replay
+# buffers.
+#
+# .. figure:: /_static/img/transforms.png
```

Comment:
this images hasn't been added to this PR and not loading:
<img width="688" alt="Screenshot 2023-11-07 at 8 57 14 AM" src="https://github.com/pytorch/tutorials/assets/5317992/53716bbb-fda4-4be6-a94f-f7205909abdb">